lib/elixir/lib/kernel/typespec.ex

```diff
@@ -533,14 +533,9 @@ defmodule Kernel.Typespec do
     module = Macro.expand(name, caller)
 
     struct =
-      if module == caller.module do
-        Module.get_attribute(module, :struct) ||
-          compile_error(caller, "struct is not defined for #{Macro.to_string(name)}")
-      else
-        module.__struct__
-      end
-
-    struct = struct |> Map.from_struct() |> Map.to_list()
+      :elixir_map.load_struct(meta, module, [], caller)
```

Comment:
I don't like that we are accessing a compiler function here and I would like to provide a proper API for this. The implementation would look like this:

```elixir
def ...(module, env) when is_atom(module) do
  :elixir_map.load_struct([line: env.line], module, [], env)
end
```

The question is: where to introduce it?

1. We could perform this expansion by default in `Macro.expand/2`. I am afraid this can be surprising and undesired in some cases.

2. We could add a specific function to `Macro`, such as `Macro.expand_struct!/2`, that does it

3. Since this is similar to `Code.ensure_compiled`, we could add `Code.ensure_struct/2`, but note that `ensure_struct` receives the environment but `ensure_compiled` does not

Thoughts?

Reply:
I like `Macro.struct!/2` because it mirrors `Kernel.struct!/1`

Reply:
I like `Macro.struct!/2` as well, though always "just" doing it on `Macro.expand/2` also seems compelling. I'm not sure...

Reply:
`Macro.struct!/2` is a bit of a cryptic name IMO, you can't really say what it's doing by its name 😕

Reply:
I would like to take a few steps back: how would users benefit from the `Macro.struct!` function (or whatever we decide)? 🤔 Maybe compile time deadlocks like the ones in the typespecs...but I guess you would also like to use functions from the module. I think you would have deadlocks either way.

On top of that, we would have two `struct!` functions, and that might be confusing.

In summary, could be an option to leave it as is and move it elsewhere if needed in the future?

Reply:
However, if we are planning to `@doc false` it, then `Macro.load_struct!` sounds good to me 🙂

Reply:
@fertapric we have ways to resolve deadlocks from functions by calling `Code.ensure_compiled!`. We don't have a way to do structs resolution though. :(

Reply:
I will go ahead with `Macro.struct!` as I don't think we will find a perfect name. I like it mirrors Kernel.